Lushu/src/main/kotlin/lushu/ContextGrammar/Grammar/Rules.kt

```diff
@@ -0,0 +1,226 @@
+     *        mutableTokens = mutableListOf("name:", "Emily", "Emily", "Emily", "is", "a", "new", "user")
+     *        index = 1
+     *        current = Node(label = "Emily", children = [Node(label = "is")])
+     * Output: (4, Node(label = "is"))
+     */
+    private fun plusCaseMatcher(inputTokens: MutableList<String>, mutableTokens: MutableList<String>, index: Int, current: Node?): Pair<Int, Node?> {
+        if(current == null)
+            return Pair(1, null)
+
+        if (isEndingPlusCase(mutableTokens[index], current)) {
+            inputTokens.clear()
+            inputTokens.addAll(mutableTokens)
+            val endPlusCaseNode = matcher(current, inputTokens[index])
+
+            return Pair(index, endPlusCaseNode)
+        }
+
+        if (current.match(mutableTokens[index])) {
+            if(current.isSensitive())
+                mutableTokens[index] = "*".repeat(mutableTokens[index].length)
+
+            return plusCaseMatcher(inputTokens, mutableTokens, index + 1, current)
+        }
+
+        return Pair(1, null)
+    }
+
+    /**
+     * match a given word against the children nodes of a current node in a tree structure.
+     * @param current The current node being evaluated.
+     * @param word The word to be matched against the children nodes.
+     * @return The matched node, if found; the terminalNode flag, if the node doesn't have children; otherwise, null.
+     * Example:
+     * Input: current = Node(label = "root", children = [Node(label = "host"), Node(label = "warning"), Node(label = "ip")]), word = "host"
+     * Output: Node(label = "host")
+     */
+    private fun matcher(current: Node?, word: String): Node?{
+        if(current == null)
+            return null
+
+        if(current!!.getChildren().isNullOrEmpty())
+            return terminalNode
+
+        return current?.getChildren()?.find { child ->
+            child.match(word)
+        }
+    }
+
+    /**
+     * Matches input tokens against a given pattern represented by a tree of nodes.
+     * @param inputTokens The list of input tokens to be matched.
+     * @param mutableTokens The mutable list of tokens used for matching and modification.
+     * @param index The current index in the inputTokens list.
+     * @param current The current node being evaluated.
+     * @return The number of tokens matched successfully.
+     * Example:
+     * Let a context "host ip: 123.123.123.123".
+     * Input:inputTokens = ["server", "host", "ip:",  "123.123.123.123", "is", "online"]
+     *       mutableTokens = ["server", "host", "ip:",  "123.123.123.123", "is", "online"]
+     *       index = 1
+     *       current = rootNode
+     * Output: 3
+     */
+    private fun matchTokensAgainstPatternContext(inputTokens: MutableList<String>, mutableTokens: MutableList<String>, index: Int, current: Node?): Int {
+        if (current == null) {
+            return 1
```

Comment:
What does `1` mean here? We should substitute these magic numbers by constants with descriptive names